core/templates/components/entity-creation-services/exploration-creation.service.spec.ts

```diff
@@ -0,0 +1,238 @@
+    ecs.explorationCreationInProgress = false;
+
+    ecs.createNewExploration();
+
+    expect(loaderService.showLoadingScreen)
+      .toHaveBeenCalledWith('Creating exploration');
+    expect(ecbas.registerNewExplorationAsync).toHaveBeenCalled();
+  });
+
+  it('should create new exploration', fakeAsync(() => {
+    spyOn(siteAnalyticsService, 'registerCreateNewExplorationEvent');
+    spyOn(urlInterpolationService, 'interpolateUrl').and.returnValue(
+      '/url/to/exp1'
+    );
+
+    expect(ecs.explorationCreationInProgress).toBe(undefined);
+    expect(windowRef.nativeWindow.location.href).toBe('');
+
+    ecs.createNewExploration();
+    tick(150);
+
+    expect(ecs.explorationCreationInProgress).toBe(true);
+    expect(windowRef.nativeWindow.location.href).toBe('/url/to/exp1');
+  }));
+
+  it('should create new exploration', fakeAsync(() => {
+    spyOn(siteAnalyticsService, 'registerCreateNewExplorationEvent');
+    spyOn(urlInterpolationService, 'interpolateUrl').and.returnValue(
+      '/url/to/exp1'
+    );
+
+    expect(ecs.explorationCreationInProgress).toBe(undefined);
+    expect(windowRef.nativeWindow.location.href).toBe('');
+
+    ecs.createNewExploration();
+    tick(150);
+
+    expect(ecs.explorationCreationInProgress).toBe(true);
+    expect(windowRef.nativeWindow.location.href).toBe('/url/to/exp1');
+  }));
+
+  it('should handle error if exploration creation fails', fakeAsync(() => {
+    spyOn(siteAnalyticsService, 'registerCreateNewExplorationEvent');
+    spyOn(urlInterpolationService, 'interpolateUrl');
+    spyOn(loaderService, 'hideLoadingScreen');
+
+    ecbas.throwError = true;
+
+    expect(ecs.explorationCreationInProgress).toBe(undefined);
+    expect(windowRef.nativeWindow.location.href).toBe('');
+
+    ecs.createNewExploration();
+    tick(150);
+
+    expect(ecs.explorationCreationInProgress).toBe(false);
+    expect(windowRef.nativeWindow.location.href).toBe('');
+    expect(siteAnalyticsService.registerCreateNewExplorationEvent)
+      .not.toHaveBeenCalled();
+    expect(urlInterpolationService.interpolateUrl).not.toHaveBeenCalled();
+    expect(loaderService.hideLoadingScreen).toHaveBeenCalled();
+  }));
+
+  it('should show upload exploration modal', fakeAsync(() => {
+    spyOn(ngbModal, 'open').and.returnValue(
+      <NgbModalRef>{
+        result: Promise.resolve({
+          yamlFile: ''
+        })
+      }
+    );
+    spyOn(csrfTokenService, 'getTokenAsync').and.resolveTo('sample-csrf-token');
+
+    // @ts-ignore in order to ignore JQuery properties that should
+    // be declared.
+    spyOn($, 'ajax').and.callFake((options) => {
+      let d = $.Deferred();
+      d.resolve(
+        options.dataFilter(')]}\',\n{"explorationId": "expId"}')
+      );
+      return d.promise();
+    });
+
+    ecs.showUploadExplorationModal();
+    tick();
+
+    expect(windowRef.nativeWindow.location.href).toBe('/create/expId');
+  }));
+
+  it('should show upload exploration modal', fakeAsync(() => {
```

Comment:
Description is unclear.
can we have this pattern `should do X if Y happens`?

Reply:
Done